File: sdk/symbol_dart_sdk/bin/symbol/Network.dart
import '../Network.dart' as basic_network;
import '../NetworkTimestamp.dart' as basic_timestamp;
import '../ByteArray.dart';
import '../CryptoTypes.dart';
import 'package:base32/base32.dart';
import 'dart:typed_data';
import 'package:pointycastle/export.dart';

class NetworkTimestamp extends basic_timestamp.NetworkTimestamp {
  NetworkTimestamp(int timestamp) : super(timestamp);

  NetworkTimestamp addMilliseconds(int count) {
    return NetworkTimestamp(timestamp + count);
  }

  @override
  NetworkTimestamp addSeconds(int count) {
    return addMilliseconds(1000 * count);
  }
}

class Address extends ByteArray {
  // Byte size of raw address.
  static const int SIZE = 24;

  // Length of encoded address string.
  static const int ENCODED_SIZE = 39;

  // Creates a Symbol address.
  Address(dynamic addressInput) : super(SIZE, _extractAddressBytes(addressInput));

  static Uint8List _extractAddressBytes(dynamic addressInput) {
    if (addressInput is String){
      return base32.decode('${addressInput}A').sublist(0, -1);
    }

    if (addressInput is Address){
      return addressInput.bytes;
    }

    return addressInput;
  }

  // Returns string representation of this object.
  @override
  String toString() {
    return base32.encode(Uint8List.fromList([...bytes, 0])).substring(0, -1);
  }
}

class Network extends basic_network.Network {
  // Symbol main network.
  static Network MAINNET = Network(
    'mainnet',
    0x68,
    DateTime.utc(2021, 2, 16, 0, 6, 25),
    Hash256('57F7DA205008026C776CB6AED843393F04CD458E0AA2D9F1D5F31A402072B2D6'),
  );

  // Symbol test network.
  static Network TESTNET = Network(
    'testnet',
    0x98,
    DateTime.utc(2022, 9, 31, 21, 7, 47),
    Hash256('49D6E1CE276A85B70EAFE52349AACCA389302E7A9754BCF1221E79494FC665A4'),
  );

  // Symbol well known networks.
  static List<Network> NETWORKS = [MAINNET, TESTNET];

  // Network generation hash seed.
  Hash256 generationHashSeed;

  // Creates a new network with the specified name, identifier byte and generation hash seed.
  Network(String name, int identifier, DateTime epochTime, Hash256 generationHashSeed)
    : generationHashSeed = generationHashSeed,
    super(
      name,
      identifier,
      basic_timestamp.NetworkTimestampDatetimeConverter(epochTime, 'milliseconds'),
      () => SHA3Digest(256),
      (Uint8List addressWithoutChecksum, Uint8List checksum) =>
          Uint8List.fromList([...addressWithoutChecksum, ...checksum.sublist(0, 3)]),
      Address,
      NetworkTimestamp,
    );
}

File: sdk/php/src/symbol/Network.php
<?php

namespace SymbolSdk\Symbol;

use SymbolSdk\BinaryData;
use SymbolSdk\Network\Network as BasicNetwork;
use SymbolSdk\Utils\Converter;
use SymbolSdk\Symbol\Address;
use SymbolSdk\Network\NetworkTimestampDatetimeConverter;
use SymbolSdk\Network\NetworkTimestamp;
use DateTime;
use SymbolSdk\CryptoTypes\Hash256;

class Network extends BasicNetwork
{
  public static $MAINNET;

  public static $TESTNET;

  public static $NETWORKS;

  public $generationHashSeed;

  const BASE32_RFC4648_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

  public function __construct(string $name, int $identifier, DateTime $epochTime, Hash256 $generationHashSeed)
  {
    parent::__construct(
      $name,
      $identifier,
      new NetworkTimestampDatetimeConverter($epochTime, 'milliseconds'),
      'sha3-256',
      function ($addressWithoutChecksum, $checksum) {
        return new Address($addressWithoutChecksum . substr($checksum, 0, 3));
      },
      Address::class,
      NetworkTimestamp::class
    );

    $this->generationHashSeed = $generationHashSeed;
  }

  /**
   * Initialize MAINNET and TESTNET networks
   */
  public static function initialize()
  {
    self::$MAINNET = new Network(
      'mainnet',
      0x68,
      new DateTime('2021-03-16T00:06:25Z'),
      new Hash256('57F7DA205008026C776CB6AED843393F04CD458E0AA2D9F1D5F31A402072B2D6')
    );

    self::$TESTNET = new Network(
      'testnet',
      0x98,
      new DateTime('2022-10-31T21:07:47Z'),
      new Hash256('49D6E1CE276A85B70EAFE52349AACCA389302E7A9754BCF1221E79494FC665A4')
    );

    self::$NETWORKS = [self::$MAINNET, self::$TESTNET];
  }

  public function isValidAddressString($addressString) {
		if ($this->_addressClass->getConstant('ENCODED_SIZE') !== strlen($addressString))
			return false;

		for ($i = 0; $i < strlen($addressString); ++$i) {
      if (strpos(self::BASE32_RFC4648_ALPHABET, $addressString[$i]) === false)
				return false;
		}
		return $this->isValidAddress($this->_addressClass->newInstance($addressString));
	}

  public function isValidAddress(BinaryData $address): bool{
    if(Converter::binaryToInt($address->binaryData[0], 1) != $this->identifier)
      return false;
    $hash = hash($this->_addressHasher, substr($address->binaryData, 0, 1 + 20 ), true);
    $checkSumFromAddress = substr($address->binaryData, 1 + 20);
    $calculatedChecksum = substr($hash, 0, strlen($checkSumFromAddress));
    for ($i = 0; $i < strlen($checkSumFromAddress); ++$i) {
			if ($checkSumFromAddress[$i] != $calculatedChecksum[$i])
				return false;
		}
		return true;
  }

  public function toDatetime(NetworkTimestamp $referenceNetworkTimestamp) {
		return $this->datetimeConverter->toDatetime($referenceNetworkTimestamp->timestamp);
	}

  public function fromDatetime(DateTime $referenceDatetime) {
		return $this->networkTimestampClass->newInstance($this->datetimeConverter->toDifference($referenceDatetime));
	}

  public function __toString()
  {
    return $this->name;
  }
}
